File: javascript/src/virtual-proxy/virtual-proxy.js
/** Creates the real subject on the first request and forwards every request to it. */
export class VirtualProxy {
  #loader;
  #realSubject;

  constructor(loader) {
    this.#loader = loader;
  }

  request() {
    this.#realSubject ??= this.#loader();
    return this.#realSubject.request();
  }

  isLoaded() {
    return this.#realSubject !== undefined;
  }
}

File: python/src/patterns/virtualproxy/virtual_proxy.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable

    from .subject import Subject


class VirtualProxy:
    """Creates the real subject on the first request and forwards every request to it."""

    def __init__(self, loader: Callable[[], Subject]) -> None:
        self._loader = loader
        self._real_subject: Subject | None = None

    def request(self) -> str:
        if self._real_subject is None:
            self._real_subject = self._loader()
        return self._real_subject.request()

    def is_loaded(self) -> bool:
        return self._real_subject is not None
